docs/getting-started/03-cli.md

```diff
@@ -90,6 +90,10 @@ curl -v http://127.0.0.1:4010/pets/123 -H "Prefer: dynamic=false"
 
 > **Remember:** you can combine `code`, `example` and `dynamic` parameters. By default, when the `code` parameter is not given, Prism will always try to fetch an example using the `exampleKey` from the 200 HTTP responses **only**. You'll need to use the `code` parameter alongside the `example` one for any specific example using a different HTTP status code than 200.
 
+#### Circular references
+
+Even though Prism is technically able to internally handle circular references, the CLI will refuse to mock the provided document in case any circular reference is detected. This is essentially because serialising a circular reference is difficult and very dependant on the content type.
```

Comment:
> Have you tried providing the whole spec as the third parameter to openapi-sampler

I don't, and it is not that easy since Prism HTTP does not see the document at all, it just deals with Operations array. We could do that but we need to have a discussion about it somehow and decide how to go.

As far as I am concerned, the less Prism does, the better it will work.

Reply:
> All of them need to have a place where the cycle is being broken because it wouldn't be possible to generate a valid payload

Yes, that is essentially the problem. We can handle circular refs in memory but there's no way we can serialise them as there are really too many variables around that (the content type being the first one in mind), although we can validate, that for sure.

I guess we can revisit this in a successive PR/discussion?

Reply:
> I guess we can revisit this in a successive PR/discussion?
For sure, the aim of this task is to be more descriptive to the user. Even if we do it wrong ;)

> As far as I am concerned, the less Prism does, the better it will work.
We should not think about it as about resolving refs. These are handled correctly by `http-spec` (`json-schema-ref-parser` or whatever). 

This is about _generating_ payload from a schema. When resolving ref, we might want to know that something is a circular structure. But when we want to generate a mocked payload, `openapi-sampler` have a freedom to say: "ok, this time parentId will be empty". This produces non-circular, serializable structure.

IMO, we should feed `openapi-sampler` with document. Which is partially sad, because it breaks the beauty of separated concerns in prism.

Reply:
> But when we want to generate a mocked payload, openapi-sampler have a freedom to say: "ok, this time parentId will be empty". This produces non-circular, serializable structure.

Oh this is very interesting. Can OpenAPI Sampler deal with circular references? As far as I know it does not but I'd be so happy to be wrong about it.

We should also verify that with json-schema-faker — do you know if that can handle self-references or circular structures?

Reply:
I've been checking that quite some time ago and as far as I remember `openapi-sampler` dealt with such situation by not filling the circular field as early as possible.

So a schema similar to `type Node = { children: Node[] }` would result in instance like `const node = { children: [] }`. Which is perfectly valid. But thats only my volatile memory. Needs to recheck.

I have no knowledge how jsf deals with that.

Reply:
I've been checking that quite some time ago and as far as I remember openapi-sampler dealt with such situation by not filling the circular field as early as possible.

So a schema similar to type Node = { children: Node[] } would result in instance like const node = { children: [] }. Which is perfectly valid. But thats only my volatile memory. Needs to recheck.

I have no knowledge how jsf deals with that.